system_metrics_collector/test/topic_statistics_collector/test_subscriber_topic_statistics.cpp

```diff
@@ -0,0 +1,406 @@
+  uint64_t getSubscriptionCount() const
+  {
+    return publisher_->get_subscription_count();
+  }
+
+private:
+  typename rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr publisher_;
+};
+
+/**
+ * Test fixture to test a SubscriberTopicStatisticsNode node
+ */
+class SubscriberTopicStatisticsNodeTestFixture : public testing::Test
+{
+public:
+  void SetUp() override
+  {
+    rclcpp::init(0, nullptr);
+
+    rclcpp::NodeOptions options;
+    options.append_parameter_override(
+      system_metrics_collector::collector_node_constants::kPublishPeriodParam,
+      kVeryLongPublishPeriod.count());
+    options.append_parameter_override(constants::kCollectStatsTopicNameParam, kTestTopicName);
+
+    imu_publisher_ = std::make_shared<ImuMessagePublisher>();
+    EXPECT_EQ(imu_publisher_->getSubscriptionCount(), 0);
+
+    test_topic_stats_node_ = std::make_shared<TestSubscriberTopicStatisticsNode>(
+      kStatsCollectorNodeName, options);
+
+    ASSERT_GT(test_topic_stats_node_->GetCollectorCount(), 0);
+
+    const auto all_collected_data = test_topic_stats_node_->GetCollectorData();
+    for (const auto & data : all_collected_data) {
+      ASSERT_TRUE(std::isnan(data.average));
+      ASSERT_TRUE(std::isnan(data.min));
+      ASSERT_TRUE(std::isnan(data.max));
+      ASSERT_TRUE(std::isnan(data.standard_deviation));
+      ASSERT_EQ(0, data.sample_count);
+    }
+  }
+
+  void TearDown() override
+  {
+    test_topic_stats_node_->shutdown();
+    EXPECT_EQ(State::PRIMARY_STATE_FINALIZED, test_topic_stats_node_->get_current_state().id());
+    EXPECT_FALSE(test_topic_stats_node_->IsPublisherActivated());
+
+    test_topic_stats_node_.reset();
+    imu_publisher_.reset();
+    rclcpp::shutdown();
+  }
+
+protected:
+  // this test is not designed to have the statistics published and reset at any point of the test,
+  // so this is defining the publish period to be something amply larger than the test duration
+  // itself
+  static constexpr std::chrono::milliseconds kVeryLongPublishPeriod = 2 * kTestDuration;
+  std::shared_ptr<TestSubscriberTopicStatisticsNode> test_topic_stats_node_;
+  std::shared_ptr<ImuMessagePublisher> imu_publisher_;
+};
+
+/**
+ * Fixture to bringup and teardown rclcpp
+ */
+class RclcppFixture : public ::testing::Test
+{
+public:
+  void SetUp() override
+  {
+    rclcpp::init(0, nullptr);
+  }
+
+  void TearDown() override
+  {
+    rclcpp::shutdown();
+  }
+};
+
+/**
+ * Generate an IMU message with a valid header value.
+  *
+  * @return shared_ptr to an IMU message
+ */
+std::shared_ptr<sensor_msgs::msg::Imu> GetImuMessageWithHeader()
+{
+  auto message = sensor_msgs::msg::Imu{};
+  message.header = std_msgs::msg::Header{};
+  message.header.stamp = rclcpp::Time{kAnyTimestamp};
+  return std::make_shared<sensor_msgs::msg::Imu>(message);
+}
+
+constexpr std::chrono::milliseconds
+SubscriberTopicStatisticsNodeTestFixture::kVeryLongPublishPeriod;
+
+TEST_F(SubscriberTopicStatisticsNodeTestFixture, TestStartAndStop) {
```

Comment:
Here and below throughout the whole file. Please use EXPECT instead of ASSERT whenever possible https://testing.googleblog.com/2008/07/tott-expect-vs-assert.html